### PR TITLE
KYLIN-4903 cache parent datasource to accelerate next layer's cuboid building

### PR DESCRIPTION
**Pull request by zhengshengjun**
…building

## Proposed changes

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.

## Types of changes

What types of changes does your code introduce to Kylin?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have create an issue on [Kylin's jira](https://issues.apache.org/jira/browse/KYLIN), and have described the bug/feature there in detail
- [x] Commit messages in my PR start with the related jira ID, like "KYLIN-0000 Make Kylin project open-source"
- [x] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If this change need a document change, I will prepare another pr against the `document` branch
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at user@kylin or dev@kylin by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: kylin-spark-project/kylin-spark-engine/src/main/scala/org/apache/kylin/engine/spark/job/CubeBuildJob.java [hunk 1]
```diff
@@ -157,7 +157,12 @@ protected void doExecute() throws Exception {
                 logger.info("Triggered cube planner phase one .");
         }
 
-        buildLayoutWithUpdate = new BuildLayoutWithUpdate();
+        if (config.isReuseParentDataSource()) {
+            buildLayerWithUpdate = new BuildLayerWithUpdate();
+        } else {
+            buildLayoutWithUpdate = new BuildLayoutWithUpdate();
```
Comment (zzcclp) [inline kylin-spark-project/kylin-spark-engine/src/main/scala/org/apache/kylin/engine/spark/job/CubeBuildJob.java:163]: Why needs these two similar Class? It's better to refactor the original class 'BuildLayoutWithUpdate' to support this feature.

File: kylin-spark-project/kylin-spark-engine/src/main/scala/org/apache/kylin/engine/spark/job/BuildLayerWithUpdate.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+            this.st = st;
+            this.seg = seg;
+            this.cuboidsRowCount = cuboidsRowCount;
+        }
+
+        public LayerJobResult build() {
+            Collection<LayoutEntity> toBuildCuboids = info.getToBuildCuboids();
+            Dataset<Row> parentDS = info.getParentDS();
+            if (toBuildCuboids.size() > 1) {
+                parentDS.cache();
```
Comment (zzcclp) [inline kylin-spark-project/kylin-spark-engine/src/main/scala/org/apache/kylin/engine/spark/job/BuildLayerWithUpdate.java:129]: How to set other storage level? It's better to add a configuration to support setting other storage level.

File: core-common/src/main/java/org/apache/kylin/common/KylinConfigBase.java [hunk 3]
```diff
@@ -3084,4 +3084,8 @@ public String getKerberosJaasConfPath() {
     public String getKerberosPrincipal() {
         return getOptional("kylin.kerberos.principal");
     }
+
+    public boolean isReuseParentDataSource() {
+        return Boolean.parseBoolean(getOptional("kylin.engine.spark.reuse.parent.rdd", FALSE));
+    }
```
Comment (zzcclp) [inline core-common/src/main/java/org/apache/kylin/common/KylinConfigBase.java:3090]: IMO, don't need this configuration, we can add a configuration named 'kylin.engine.spark.parent-dataset.storage.level' and set the storage level directly, if the storage level is 'NONE', it means it doesn't need to cache parent dataset.

File: kylin-spark-project/kylin-spark-engine/src/main/scala/org/apache/kylin/engine/spark/job/CubeBuildJob.java [hunk 4]
```diff
@@ -319,45 +324,61 @@ private void build(Collection<NBuildSourceInfo> buildSourceInfos, SegmentInfo se
...
+            buildLayoutWithUpdate.updateLayout(seg, config);
         }
 
-        infos.recordCuboidsNumPerLayer(seg.id(), cuboidsNumInLayer);
-        buildLayoutWithUpdate.updateLayout(seg, config);
-
         // decided the next layer by current layer's all indexes.
         st.decideTheNextLayer(allIndexesInCurrentLayer, seg);
         return constructTheNextLayerBuildInfos(st, seg, allIndexesInCurrentLayer);
+
```
Comment (zzcclp) [inline kylin-spark-project/kylin-spark-engine/src/main/scala/org/apache/kylin/engine/spark/job/CubeBuildJob.java:381]: remove this blank line.

File: core-common/src/main/java/org/apache/kylin/common/KylinConfigBase.java (hunk 3 above)
Comment (zhengshengjun) [reply core-common/src/main/java/org/apache/kylin/common/KylinConfigBase.java:3090]: thanks,  good advice indeed ~

File: kylin-spark-project/kylin-spark-engine/src/main/scala/org/apache/kylin/engine/spark/job/CubeBuildJob.java [hunk 5]
```diff
@@ -319,45 +324,61 @@ private void build(Collection<NBuildSourceInfo> buildSourceInfos, SegmentInfo se
...
                     toBuildCuboids.stream().map(LayoutEntity::getId).collect(Collectors.toList()));
             cuboidsNumInLayer += toBuildCuboids.size();
             Preconditions.checkState(!toBuildCuboids.isEmpty(), "To be built cuboids is empty.");
-            Dataset<Row> parentDS = info.getParentDS();
-            // record the source count of flat table
-            if (info.getLayoutId() == ParentSourceChooser.FLAT_TABLE_FLAG()) {
-                cuboidsRowCount.putIfAbsent(info.getLayoutId(), parentDS.count());
-            }
+            info.getToBuildCuboids().stream().forEach(allIndexesInCurrentLayer::add);
+            infos.recordCuboidsNumPerLayer(seg.id(), cuboidsNumInLayer);
```
Comment (zzcclp) [inline kylin-spark-project/kylin-spark-engine/src/main/scala/org/apache/kylin/engine/spark/job/CubeBuildJob.java:339]: the value of 'cuboidsNumInLayer' is the total count of current segments, why needs to set this value into infos per loop

**Comment by zhengshengjun** [comment]
thanks for @zzcclp's advice, i'll refine the implementation.

**Comment by zhengshengjun** [comment]
@zzcclp updated, please help me to review this again  ^_^

File: kylin-spark-project/kylin-spark-engine/src/main/scala/org/apache/kylin/engine/spark/job/BuildLayoutWithUpdate.java [hunk 6]
```diff
@@ -38,8 +46,45 @@
...
+    }
+
+    public void cacheAndRegister(long layoutId, Dataset<Row> dataset) throws InterruptedException{
+        if (!persistParentDataset) {
+            return;
+        }
+        logger.info("persist dataset of layout: {}", layoutId);
+        semaphore.acquire();
+        layout2DataSet.put(layoutId, dataset);
+        dataset.persist(StorageLevel.fromString(kylinConfig.getParentDatasetStorageLevel()));
```
Comment (zzcclp) [inline kylin-spark-project/kylin-spark-engine/src/main/scala/org/apache/kylin/engine/spark/job/BuildLayoutWithUpdate.java:70]: Get the value of "kylinConfig.getParentDatasetStorageLevel()" in the BuildLayoutWithUpdate initialization method, and then check the validity of the value.

File: kylin-spark-project/kylin-spark-engine/src/main/scala/org/apache/kylin/engine/spark/job/CubeBuildJob.java [hunk 7]
```diff
@@ -330,6 +330,11 @@ private void build(Collection<NBuildSourceInfo> buildSourceInfos, SegmentInfo se
             cuboidsNumInLayer += toBuildCuboids.size();
             Preconditions.checkState(!toBuildCuboids.isEmpty(), "To be built cuboids is empty.");
             Dataset<Row> parentDS = info.getParentDS();
+
+            if (toBuildCuboids.size() >= 1) {
```
Comment (zzcclp) [inline kylin-spark-project/kylin-spark-engine/src/main/scala/org/apache/kylin/engine/spark/job/CubeBuildJob.java:334]: why is '>='? if the size of toBuildCuboids is 1, it doesn't need to cache parent dataset?

Comment (zhengshengjun) [reply kylin-spark-project/kylin-spark-engine/src/main/scala/org/apache/kylin/engine/spark/job/CubeBuildJob.java]: you are right, fixed.

File: kylin-spark-project/kylin-spark-engine/src/main/scala/org/apache/kylin/engine/spark/job/BuildLayoutWithUpdate.java [hunk 8]
```diff
@@ -38,8 +46,45 @@
...
     private int currentLayoutsNum = 0;
+    private Map<Long, AtomicLong> toBuildCuboidSize = new ConcurrentHashMap<>();
+    private Semaphore semaphore;
+    private Map<Long, Dataset<Row>> layout2DataSet = new ConcurrentHashMap<>();
+    private KylinConfig kylinConfig;
+    private boolean persistParentDataset;
+
+    public BuildLayoutWithUpdate(KylinConfig kylinConfig) {
+        this.kylinConfig = kylinConfig;
+        this.persistParentDataset = !kylinConfig.getParentDatasetStorageLevel().equals(StorageLevel.NONE());
```
[the remark this answers is not in the document]
Comment (zhengshengjun) [reply kylin-spark-project/kylin-spark-engine/src/main/scala/org/apache/kylin/engine/spark/job/BuildLayoutWithUpdate.java]: It's my fault ,thanks for finding it ~

File: kylin-spark-project/kylin-spark-engine/src/main/scala/org/apache/kylin/engine/spark/job/BuildLayoutWithUpdate.java (hunk 6 above)
Comment (zhengshengjun) [reply kylin-spark-project/kylin-spark-engine/src/main/scala/org/apache/kylin/engine/spark/job/BuildLayoutWithUpdate.java]: good suggestion

**Comment by zzcclp** [comment]
LGTM

**Pull request merged**